api/client/firewallrules/client_test.go

```diff
@@ -4,140 +4,114 @@
 package firewallrules_test
 
 import (
+	"github.com/golang/mock/gomock"
 	"github.com/juju/errors"
 	jc "github.com/juju/testing/checkers"
 	gc "gopkg.in/check.v1"
 
-	basetesting "github.com/juju/juju/api/base/testing"
+	basemocks "github.com/juju/juju/api/base/mocks"
 	"github.com/juju/juju/api/client/firewallrules"
 	apiservererrors "github.com/juju/juju/apiserver/errors"
 	"github.com/juju/juju/rpc/params"
-	"github.com/juju/juju/testing"
 )
 
 type FirewallRulesSuite struct {
-	testing.BaseSuite
 }
 
 var _ = gc.Suite(&FirewallRulesSuite{})
 
 func (s *FirewallRulesSuite) TestSetFirewallRule(c *gc.C) {
-	apiCaller := basetesting.APICallerFunc(
-		func(objType string,
-			version int,
-			id, request string,
-			a, result interface{},
-		) error {
-			c.Check(objType, gc.Equals, "FirewallRules")
-			c.Check(id, gc.Equals, "")
-			c.Check(request, gc.Equals, "SetFirewallRules")
-
-			args, ok := a.(params.FirewallRuleArgs)
-			c.Assert(ok, jc.IsTrue)
-			c.Assert(args.Args, gc.HasLen, 1)
-
-			rule := args.Args[0]
-			c.Assert(rule.KnownService, gc.Equals, params.SSHRule)
-			c.Assert(rule.WhitelistCIDRS, jc.DeepEquals, []string{"192.168.1.0/32"})
-
-			if results, ok := result.(*params.ErrorResults); ok {
-				results.Results = []params.ErrorResult{{
-					Error: apiservererrors.ServerError(errors.New("fail"))}}
-			}
-			return nil
-		})
-
-	client := firewallrules.NewClient(apiCaller)
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+
+	args := params.FirewallRuleArgs{
+		Args: []params.FirewallRule{{
+			KnownService:   "ssh",
+			WhitelistCIDRS: []string{"192.168.1.0/32"},
+		}},
+	}
+	res := new(params.ErrorResults)
+	results := params.ErrorResults{
+		Results: []params.ErrorResult{{
+			Error: apiservererrors.ServerError(errors.New("fail"))}},
+	}
+
+	mockFacadeCaller := basemocks.NewMockFacadeCaller(ctrl)
+	mockFacadeCaller.EXPECT().FacadeCall("SetFirewallRules", args, res).SetArg(2, results).Return(nil)
+	client := firewallrules.NewClientFromCaller(mockFacadeCaller)
+
 	err := client.SetFirewallRule("ssh", []string{"192.168.1.0/32"})
 	c.Assert(err, gc.ErrorMatches, "fail")
 }
 
 func (s *FirewallRulesSuite) TestSetFirewallRuleFacadeCallError(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+
 	msg := "facade failure"
-	apiCaller := basetesting.APICallerFunc(
-		func(objType string,
-			version int,
-			id, request string,
-			a, result interface{},
-		) error {
-			c.Check(objType, gc.Equals, "FirewallRules")
-			c.Check(id, gc.Equals, "")
-			c.Check(request, gc.Equals, "SetFirewallRules")
-			return errors.New(msg)
-		})
-	client := firewallrules.NewClient(apiCaller)
+	args := params.FirewallRuleArgs{
+		Args: []params.FirewallRule{{
+			KnownService: "ssh",
+		}},
+	}
+	res := new(params.ErrorResults)
+	results := params.ErrorResults{
+		Results: []params.ErrorResult{{
+			Error: apiservererrors.ServerError(errors.New("fail"))}},
```

Comment:
this part will never be evaluated by the client. best to remove and not confusing the test intention.